modules/programs/afew.nix

```diff
@@ -6,21 +6,86 @@ let
 
   cfg = config.programs.afew;
 
+  mkIniKeyValue = key: value:
+    let
+      tweakVal = v:
+        if isString v then v
+        else if isList v then concatMapStringsSep ";" tweakVal v
+        else if isBool v then (if v then "true" else "false")
+        else toString v;
+    in
+      "${key}=${tweakVal value}";
+
+  /* Return a list of name value pairs where name is "class.index" and value is
+     an attribute set of arbitrary afew options.
+
+     Example:
+       filters = [
+            { class="Class", q="a"; }
+            { q = "b"}
+          ]
+      => [{ name = "Class.1"; value = { "q" = "a"; }; } { name = "Filter.2"; value = { "q" = "b"; }; }]
+  */
+  processAfewFilters = afewFilters:
+    let
+      getClass = attrs:
+        attrByPath ["class"] "Filter" attrs;
+      removeClass = attrs:
+        filterAttrs (n: v: n != "class") attrs;
+      indexed = xs:
+        imap0 (i: v: v // { class = "${getClass v}.${toString (i+1000)}"; }) xs;
+      toPair = mkKey: mkValue: attrs:
+        nameValuePair (mkKey attrs) (mkValue attrs);
+    in
+      map (toPair getClass removeClass) (indexed afewFilters);
+
+  /* Convert a list of name value pairs to a list of INI sections */
+  toOrderedINI = {
+    mkKeyValue ? mkKeyValueDefault {} "="
+  }: afewFilters:
+    let
+      mkSection = pair: ''
+        ${mkSectionName pair}
+        ${toKV pair.value}'';
+      mkSectionName = pair: "[${pair.name}]";
+      toKV = attrs: generators.toKeyValue { inherit mkKeyValue; } attrs;
+    in
+      concatStringsSep "\n" (map mkSection afewFilters);
+
 in
 
 {
   options.programs.afew = {
     enable = mkEnableOption "the afew initial tagging script for Notmuch";
 
+    filters = mkOption {
```

Comment:
I find this a bit tricky to follow. Would it be possible to express it with a number of separate options instead? Like `programs.afew.filters.spam`, `programs.afew.filters.killThreads`, etc.

So the example given for this option would instead be
```
programs.afew = {
  filters = {
    spam = [
      {
        query = "frompointyheaded@boss.com";
        tags = ["-new" "+boss"];
        message = "Message from above";
      }
    ];

    inbox = {};
  };
}
```

Reply:
Thanks for the input :)

In your example, the "frompointyheaded@boss.com" filter is a generic filter (not part of the spam filter), and there maybe any number of generic filters like this one. So it would be more like:

```
programs.afew = {
  filters = {
    spam = {};

    filter = {
      query = "frompointyheaded@boss.com";
      tags = ["-new" "+boss"];
      message = "Message from above";
    };

    filter = {
      query = "somebodyelse@boss.com";
      tags = ["-new" "+boss"];
      message = "Another message from above";
    };

    inbox = {};
  };
}
```

Unfortunately, filters in a few are applied in order which introduces dependencies. As I understand, attribute sets in nix are ordered alphabetically, so `filters` needs to be a list. I'm very new to nix so I could be wrong on this one :) I'm happy to change this PR if you know of a way to bend the way nix orders attributes in an attribute set.